### ISSUE TITLE
Ensure that Kamon becomes test friendly

### ISSUE THREAD

**Issue by ivantopo** [kamon-io/Kamon#202]
In the current scheme users need to start Kamon for everything to work properly and at this point it includes tests, even while users might not want/need to have Kamon on while they are running tests. At the moment, I see that we need to support the following cases:

* No Kamon at all. By this I mean, no need to start Kamon and simply let the application run, even while
Kamon is not started.
* Metrics and traces running, no reporters. Sometimes users might want to measure or trace some things as part of their tests, probably to ensure that they are correctly propagating a TraceContext if they do it manually or that certain pieces of code are actually being measured. Support for this case might even be useful in our own tests.
* Full Kamon running, with reporting modules and all. This is just about doing the same thing you do for the regular application, but it would be nice to document how people could easily ensure that Kamon is started/shutdown when needed.

for reference: https://groups.google.com/forum/#!topic/kamon-user/hnXav2WtFZ8

**Comment by kodemaniak** [comment]
+1

I think the most important case here is the first one. If I would be able to just turn off/disable/not start Kamon during tests, that would be extremely helpful already. BTW, if you exclusively use the built-in metrics (actor metrics, dispatcher metrics) that already works. The issue only seems to exist with the custom metrics.

However, additionally having means to check that the metrics are correctly collected, or that traces work as expected, would be a nice addition. Maybe sth. like a TestReporter that allows for assertions on different aspects of the metrics collection would be a way to go.

I am not sure, though, whether it is easy right now to disable reporters via configuration. My build, e.g., contains my custom OpenTSDB reporter as a dependency. If I start Kamon during a test, I assume that currently it would also load the reporter and would require a running OpenTSDB. So that should be configurable.

**Comment by barnone** [comment]
+1 
fyi..this is chris. I reported this issue in the user forum.

My thought is that by default, you could have the current behavior which enforces Kamon.start(), however this behavior could be silenced by adding a config setting similar to the weaver warning setting. When enforcement is off, Kamon.start would not be enforced and calling Kamon APIs would not throw exceptions that Kamon is not started, API calls would instead be a noOp (do nothing) if Kamon.start() had not yet been called.

I do agree that it would also be useful to test things like context propagation in unit tests. For example we unit test our actors individually and also integrated. It would be great to be able to use Kamon.start() Kamon.shutdown() to frame where you want Kamon to come alive while still being able to test other code that uses Kamon where you do not want the system active. However, this capability is definitely much lower priority to me than the first use case above.

**Comment by ancane** [comment]
+1 for this feature.

**Comment by tobnee** [comment]
+1 just ran into this myself

**Comment by mgillis** [comment]
+1

**Comment by moliware** [comment]
+1

**Comment by lofifnc** [comment]
+1

**Comment by pimeys** [comment]
+1

**Comment by ivantopo** [comment]
Dear folks! 

We just did a few minor changes that might work to solve this issue, or at least alleviate the pain generated by it. The requirement of calling `Kamon.start(...)` will remain, but with the changes introduced here will allow you to bypass that requirement if desired, with a little trade off to make: If the `kamon.auto-start` system property is set to `true`, then Kamon wont throw an error if you tried to use it before starting it, but rather will try to start using the default configuration. This means that, when this system property is turned on you wont be able to pass a custom Config object for Kamon to be initialized with it. If you need to customize anything related to Kamon, then use the application.conf file in your resources folder.

Also, given that now Kamon auto-start can be triggered by any piece of code, even code that might run before you actually call `Kamon.start(...)` if you are currently doing so, a warning of duplicate calls to start will be printed (currently it is just a println but we need to log it properly, you get the idea anyway)... if that warning annoys you, you can use the `kamon.disable-duplicate-start-warning=true` system property. Please note that this properties are JVM properties that are read before we even try to initialize the typesafe-config library, so the must be passed as JVM -D parameters, not in configuration files.

You can simply add `javaOptions in Test := Seq("-Dkamon.auto-start=true")` to your sbt project and see how it goes.

This solution is far from perfect, but this problem has been around for too long and we need to do something... this is our first attempt! Please give it a try using snapshots with version `0.4.1-6897257f93b1bfc07557d2c62907429d46f4898a` from our snapshots repo (http://snapshots.kamon.io/) and let us know how it goes!

**Comment by crypticmind** [comment]
I tried this snapshot and it effectively relieved me from having to start Kamon myself (and requiring every test to mix in a trait for that, yadda yadda). I'd like to see more of an option like `-Dkamon.be-dummy-and-dont-bother-me-during-my-testing`, but in the mean time, this is a nice start!

On a final note, check the message you output in case `auto-start` hasn't been set and Kamon hasn't been started. The property must be set to true, not just set (`-Dkamon.auto-start=true`). Also mind a missing whitespace and a typo there.

**Comment by ivantopo** [comment]
@crypticmind, nice feedback! I thought about adding a test-related name for this property, but in reality this is something that affects Kamon as a whole, not just for tests.. in theory, people could be using this in their real application deployments, still, that isn't something we would like to encourage people to do, but certainly can happen.

Thanks for pointing out the spelling issues! (who is the dense motherfucker now? :D)

**Comment by helmbold** [comment]
+1 for better test support

**Comment by gmalouf** [comment]
I assume this is not in the 0.5 version that was releases recently?

**Comment by ivantopo** [comment]
Folks, the changes introduced by 026d7ae4ea90a5de74ecb91f9ebdd297f6987cb2 have already been merged to master and will go out with Kamon 0.5.1 that should be published this weekend. I opened a new issue to track improvements over the current solution as it is obviously less than ideal, please head to #248 to leave comments and feedback!

**Comment by ivantopo** [comment]
@gmalouf no, it didn't but 0.5.1 is going to be out pretty soon, stay tuned!
